### PR TITLE
incr.comp.: Use DefPathHash-based DepNodes in the serialized DepGraph and remove obsolete DefIdDirectory

### PR DESCRIPTION
**Pull request by michaelwoerister**
With this PR we don't store the dep-graph as a set of `DepNode<IndexIntoDefIdDirectory>` anymore but instead as a set of `DepNode<DefPathHash>`. Since a `DefPathHash` is a global identifier that is valid across compilation sessions, we don't need the `DefIdDirectory` anymore. 

Since a `DepNode<DefPathHash>` is bigger than a `DepNode<IndexIntoDefIdDirectory>` and our on-disk encoding of the dep-graph is inefficient, this PR will probably increase the amount of space the dep-graph takes up on disk. I'm in the process of gathering some performance data.

The changes in here are a step towards implementing ICH-based `DepNodes` (#42294).

r? @nikomatsakis

### CODE REVIEW

**Comment by michaelwoerister** [comment]
Some numbers for compiling `ripgrep`:

### New implementation
* initial runs = 24.81s 24.85s 24.72s
* re-runs      = 4.97s 4.91s 4.88s
* rg-depgraph = 6.0 MB
* grep-depgraph = 2.2 MB

### Current implementation
* initial run = 24.82s 24.98s 24.88s
* re-run      = 4.98s 4.87s 4.91s
* rg-depgraph = 4.0 MB
* grep-depgraph = 1.1 MB

I'd say the timings are pretty much equivalent. The on-disk graph gets bigger though. I'll look into a more efficient on-disk representation (which should be pretty similar to what we soon want to do anyway).

**Comment by michaelwoerister** [comment]
One more thing: This change slightly regresses the output of `Zincremental-info` because it will show `incremental: module XYZ is dirty because DepNode::Hir(<some opaque hash value>) changed or was removed` in the case of an item having been removed. For things that just have changed, you still get to see the `DefPath` and for removed things you at least get to see the *kind* of `DepNode`.

This could be fixed by storing some additional information with the DepGraph in case `-Zincremental-info` or `-Zquery-dep-graph` is passed. But I don't think it's an urgent problem.

**Comment by michaelwoerister** [comment]
OK, I have a proof of concept implementation of a new dep-graph encoding:

### New DepGraph Encoding
* initial runs = 24.55s 24.75s 24.76s
* re-runs = 4.68s 4.67s 4.71s
* rg-depgraph = 2.4 MB
* grep-depgraph = 1.1 MB

Looks like that slightly improves re-build times and uses less space than before the PR. I'll clean it up and push it.

**Comment by michaelwoerister** [comment]
Ready for review.

File: src/librustc/ty/context.rs [hunk 1]
```diff
@@ -676,6 +680,40 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         let max_cnum = s.cstore.crates().iter().map(|c| c.as_usize()).max().unwrap_or(0);
         let mut providers = IndexVec::from_elem_n(extern_providers, max_cnum + 1);
         providers[LOCAL_CRATE] = local_providers;
+
+        let def_path_hash_to_def_id = if s.opts.build_dep_graph() {
```
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: seems like a good place for a comment kind of explaining roughly what this table is for -- but a question: does this include only foreign crates? I don't see where it adds `LOCAL_CRATE` to the table...

**Review by nikomatsakis — APPROVED** [review]
I'm a bit confused how `LOCAL_CRATE` works, but presumably it does, and I just have to reskim the code. This looks nice. 

We should discuss whether we expect `DepNode<N>` to remain, or be subsumed into `QueryKey`, etc.

File: src/librustc/ty/context.rs (hunk 1 above)
Comment (michaelwoerister) [inline src/librustc/ty/context.rs]: The local crate is added below via `chain()`

**Pull request merged**